tools/android/packaging/xbmc/src/org/xbmc/kodi/XBMCJsonRPC.java.in

```diff
@@ -0,0 +1,647 @@
+        e.printStackTrace();
+        return null;
+      }
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to read JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public JSONObject request_object(String jsonRequest)
+  {
+    try
+    {
+      JSONObject resp = new JSONObject(request_string(jsonRequest));
+      return resp;
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to parse JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public JSONArray request_array(String jsonRequest)
+  {
+    try
+    {
+      JSONArray resp = new JSONArray(request_string(jsonRequest));
+      return resp;
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to parse JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public Bitmap getBitmap(String src)
+  {
+    if (src.isEmpty())
+      return null;
+
+    try
+    {
+      JSONObject req = request_object("{\"jsonrpc\": \"2.0\", \"method\": \"Files.PrepareDownload\", \"params\": { \"path\": \""
+          + src + "\"}, \"id\": \"1\"}");
+      if (req == null)
+        return null;
+
+      JSONObject result = req.getJSONObject("result");
+      String surl = result.getJSONObject("details").getString("path");
+
+      URL url = new URL(m_jsonURL + "/" + surl);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setDoInput(true);
+      connection.connect();
+      InputStream input = connection.getInputStream();
+      Bitmap myBitmap = BitmapFactory.decodeStream(input);
+      return myBitmap;
+    } catch (Exception e)
+    {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public String getBitmapUrl(String src)
+  {
+    if (src.isEmpty())
+      return null;
+
+    try
+    {
+      JSONObject req = request_object("{\"jsonrpc\": \"2.0\", \"method\": \"Files.PrepareDownload\", \"params\": { \"path\": \""
+          + src + "\"}, \"id\": \"1\"}");
+      if (req == null)
+        return null;
+
+      JSONObject result = req.getJSONObject("result");
+      String surl = result.getJSONObject("details").getString("path");
+
+      return (m_jsonURL + "/" + surl);
+    } catch (Exception e)
+    {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public Cursor search(String query)
+  {
+      String[] menuCols = new String[] {
+              BaseColumns._ID,
+              COLUMN_TITLE,
+              COLUMN_TAGLINE,
+              COLUMN_THUMB,
+              COLUMN_FANART,
+      };
+      MatrixCursor mc = new MatrixCursor(menuCols);
+
+      try
+      {
+        JSONObject req = request_object(String.format(SEARCH_MOVIES_JSON, /*"\"operator\": \"contains\", \"field\": \"title\", \"value\": \"" + query + "\"", limit));*/
+        "\"or\": [" +
+        "{\"operator\": \"contains\", \"field\": \"title\", \"value\": \"" + query + "\"}," +
+        "{\"operator\": \"contains\", \"field\": \"set\", \"value\": \"" + query + "\"}," +
```

Comment:
@Razzeee  Why do you need JSON at all to communicate with Kodi from inside Kodi ?

Asking a user to activate web server to open it's Kodi to outside world (Knowing how unsecure webserver is by nature) to have TV recommendation on the same device that Kodi is installed sounds not cool. (Specially without a default password set)

TCP 9090 on localhost is on by default and getting the port is only eventually adding a map in JNI. Same for thumb directory. Since same process no security problem at all.

Furthermore adding 2 or 3 required JNI functions to manage this would be way more efficient than JSON encoding / decoding that is quite slow on low end devices.